bin/bout-v5-macro-upgrader.py

```diff
@@ -0,0 +1,227 @@
+    for header in headers:
+        if (
+            re.search(r'^#\s*include.*(<|"){}(>|")'.format(header), source, flags=re.M)
+            is not None
+        ):
+            return source
+
+    # If the old macro isn't in the file, we can skip this fix
+    if re.search(r"\b{}\b".format(old), source) is None:
+        return source
+
+    # Now we want to find a suitable place to stick the new include
+    # Good candidates are includes of BOUT++ headers
+    includes = []
+    source_lines = source.splitlines()
+    for linenumber, line in enumerate(source_lines):
+        if re.match(r"^#\s*include.*bout/", line):
+            includes.append(linenumber)
+        if re.match(r"^#\s*include.*physicsmodel", line):
+            includes.append(linenumber)
+
+    if includes:
+        last_include = includes[-1] + 1
+    else:
+        # No suitable includes, so just stick at the top of the file
+        last_include = 0
+    source_lines.insert(last_include, '#include "{}"'.format(headers[0]))
+
+    return "\n".join(source_lines)
+
+
+def fix_ifdefs(old, source):
+    """Remove any #ifdef/#endif pairs that use the old macro
+    """
+    source_lines = source.splitlines()
+
+    in_ifdef = None
+    lines_to_pop = []
+    for linenumber, line in enumerate(source_lines):
+        if_def = re.match(r"#\s*ifdef\s*(.*)", line)
+        endif = re.match(r"#\s*endif", line)
+        if not (if_def or endif):
+            continue
+        # Now we need to keep track of whether we're inside an
+        # interesting #ifdef, as they might be nested, and we want to
+        # find the matching #endif
+        if endif:
+            if in_ifdef is not None:
+                in_ifdef -= 1
+            if in_ifdef == 0:
+                in_ifdef = None
+                lines_to_pop.append(linenumber)
+            continue
+        if if_def.group(1) == old:
+            in_ifdef = 1
+            lines_to_pop.append(linenumber)
+        elif in_ifdef is not None:
+            in_ifdef += 1
+
+    # Go over the source lines in reverse so that we don't need to
+    # recompute indices
+    for line in reversed(lines_to_pop):
+        del source_lines[line]
+
+    return "\n".join(source_lines)
+
+
+def fix_replacement(old, new, source):
+    """Straight replacements
+    """
+    return re.sub(r'([^"])\b{}\b([^"])'.format(old), r"\1{}\2".format(new), source)
+
+
+def apply_fixes(replacements, source):
+    """Apply all fixes in this module
+    """
+    modified = copy.deepcopy(source)
+
+    for replacement in replacements:
+        modified = fix_include_version_header(
+            replacement["old"], replacement["headers"], modified
+        )
+        modified = fix_ifdefs(replacement["old"], modified)
+        modified = fix_replacement(replacement["old"], replacement["new"], modified)
+
+    return modified
+
+
+def yes_or_no(question):
+    """Convert user input from yes/no variations to True/False
+
```

Comment:
Being picky, but should this blank line be deleted?

Reply:
Probably, but I'm going to pull out some these repeated functions into a `v5_upgrader` module, as I keep copy-pasting between the upgrader scripts

At some point, there will be a super "do everything" script that just calls all the `apply_fixes` functions